### ISSUE TITLE
IDE doesn't add newline at end of file

### ISSUE THREAD

**Issue by brianlheim** [supercollider/supercollider#4430]
## Expected vs. actual behavior

The IDE doesn't add a newline at the end of a file when saving. As a result, many people accidentally commit .sc/.scd files without a closing newline, which results in git complaining and time wasted in reviews. See for example https://github.com/supercollider/supercollider/pull/4374.

The IDE should at ensure any saved files have a closing newline. I haven't looked at the relevant code yet so I don't know where that check should happen. However, I would want to be sure that any implementation of this interacts well with how the IDE displays the text, including line numbers for instance.

**Comment by brianlheim** [comment]
I know at least one other IDE I use implicitly considers lines to always end with a newline. Unlike the SC IDE, which adds an extra line number if the file ends in `\n`, that IDE would only display a line number for an empty line if the file ends in `\n\n`. That seems like the correct solution to me.

**Comment by snappizz** [comment]
> See for example #4374.

i didn't edit #4374 with SCIDE though. i used to configure my editor to automatically add newline at EOL and automatically remove trailing spaces. but then i got comments like [this one](https://github.com/supercollider/supercollider/pull/4041#discussion_r214501177) and turned off such behaviors since they add commit noise.

**Comment by brianlheim** [comment]
> i didn't edit #4374 with SCIDE though. 

sorry, my mistake in assuming!

> i used to configure my editor to automatically add newline at EOF and automatically remove trailing spaces.

are these behaviors are both controlled by the same option?

> do you have a preference on whether i should just be more careful about watching for these mistakes before committing, or let the editor take care of it for me and ask reviewers to put up with some unintentional whitespace fixes?

now that we've removed trailing whitespace in most of the C++ files the latter shouldn't be as much of a problem anymore. however, if you enable the default git pre-commit hook (`mv .git/hooks/pre-commit.sample .git/hooks/pre-commit`), it will prevent you from committing a file with 'whitespace errors' (no closing newline, trailing ws, etc.) altogether.

**Comment by snappizz** [comment]
> are these behaviors are both controlled by the same option?

nope, separate, but i turned both of them off since they can create unintentional changes when editing a file. the same argument applies to trailing spaces.

> now that we've removed trailing whitespace in most of the C++ files the latter shouldn't be as much of a problem anymore.

my question is about sclang files, not C++ files.

**Comment by brianlheim** [comment]
i usually don't mind if someone adds a closing newline to a file, it's not as disruptive as the other whitespace changes.

**Comment by snappizz** [comment]
ok cool, sorry for going off topic here

**Comment by brianlheim** [comment]
no worries!
